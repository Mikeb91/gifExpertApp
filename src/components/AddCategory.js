import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ({ setCategories }) => { //Recordar que aquí se reciben los Props,en este caso de estestructura para obtener la función directamente

    const [inputValue, setInputValue] = useState(""); //Si se deja úncamente useState(), enviará un error ya que el valor del estado inicial es undefined, luego nosotros lo estamos 
                                                    //definiendo. Esto no debería pasar. Por esta razón se debe "inicializar" como string vacío. 

    const handleInputChange = (e) => {
        //console.log(e.target.value);
        setInputValue(e.target.value);
    }

    const handleSubmit = (e) => {
        e.preventDefault(); //Previene la actualización automática del navegador que hace el form justo al hacer submit (presionar enter). 
        if ( inputValue.trim().length > 2 ){
            setCategories(categs => [...categs, inputValue]);
            setInputValue('');
        }
        
    }


    return (
        <form onSubmit={handleSubmit}>
            <input
                type="text"
                value={inputValue}
                onChange={handleInputChange}
            />
        </form>
    )
}

AddCategory.propTypes = { //Regla para que el componente por obligación reciba una función en los props. 
    setCategories: PropTypes.func.isRequired
}
