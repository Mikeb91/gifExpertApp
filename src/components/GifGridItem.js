import React from 'react';
import PropTypes from 'prop-types';
import urlPropType from 'url-prop-type'; // permite validar no solamente si hay un String sino que valida si es una URL válida

export const GifGridItem = ({ title, url }) => {
    //console.log(id, title, url);
    return (
        <div className = "card animate__animated animate__fadeIn"> {/* //Como class es una palabra reservada de JS se dejó la palabra className para lo que corresponde a clases de CSS */}
            <img src={url} alt={title}></img>
            <p>{title}</p>
        </div>
    )
}

GifGridItem.propTypes = {
    title: PropTypes.string.isRequired,
    url: urlPropType.isRequired
}