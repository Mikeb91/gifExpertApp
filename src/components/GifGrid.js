import React, { useState, useEffect } from 'react';
import { GifGridItem } from './GifGridItem';

export const GifGrid = ({ category }) => {


    const [images, setImages] = useState([]);
    useEffect(() => {
        getGifs();
    }, []);

    const getGifs = async () => {
        const url = 'http://api.giphy.com/v1/gifs/search?q=goku&limit=10&api_key=WuM4HJU5aWmiRZ9mkmmSLU1xHqAZAck0';
        const resp = await fetch(url);
        const { data } = await resp.json();

        const gifs = data.map(img => {
            return {
                id: img.id,
                title: img.title,
                url: img.images?.downsized_medium.url
            }
        })
        //console.log(data);
        setImages(gifs);
    }

    return (
        <>
            <h3>{category}</h3>
            {
                /* images.map((image) =>{
                    return(<li key={image.id}>{image.title}</li>); //Se puede reemplazar por el return implícito y de paso usar desestructuración: 
                }) */

                /*  images.map(({id, title}) => (
                     <li key={id}>{title}</li>
                 )) */
                // Se reemplaza esta funcionalidad con un componente aparte: 

                images.map(img => (
                    <GifGridItem
                    key= {img.id} 
                    {...img} // AL hacer esto se envía cada una de las propiedades de cada imagen como un elemento de los props
                    />
                ))

            }

        </>
    )
}
