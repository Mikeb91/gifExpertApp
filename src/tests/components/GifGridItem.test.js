import { shallow } from "enzyme";
import { GifGridItem } from "../../components/GifGridItem";

describe('Pruebas en GifGridItem', () => {

    const title = 'Un título';
    const url = 'https://localhost/img.jpg';
    
    let wrapper = shallow(<GifGridItem
        title={title}
        url = {url}
    />);
    beforeEach(() => {
        wrapper = shallow(<GifGridItem
            title={title}
            url = {url}
        />);
    })

    test('Should show the component correctly', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should has a parragraph including title', () => {
        const p = wrapper.find('p');
        expect(p.text().trim()).toBe(title);
    })
    
    test('should get the img same as url and alt of props', () => {
        const img = wrapper.find('img');  // Si hay múltiples en nuestra vista, se puede utilziar id con #, o identificador de clase con . o si son varias con .at()
        //console.log(img.html()) //Si utilizamos img.html se mostrará en consola el String de la etiqueta incluyendo la info de propiedades
        console.log(img.props()) //En cambio si se utiliza .props() se obtendrá un objeto con las propiedades del componente. Si necesitamos hacer test con cada propiedad se puede
                                 //obtener de esta manera. 

        expect(img.props().src).toBe(url);
        expect(img.props().alt).toBe(title);

    })

    test('should has the class animate__fadeIn', () => {
        const requiredClassName = 'animate__fadeIn';
        const div = wrapper.find('div');
        const className = div.props().className;
        expect(className == null).toBe(false); //Se valida primero que la propiedad exista, es decir, que el div tenga clases asignadas
        if(className != null){  //Posteriormente se valida, en caso de que si exista la propiedad, que esta contenga la clase obligatoria
            expect(className.includes(requiredClassName)).toBe(true);
        }
    })
    
    

})
