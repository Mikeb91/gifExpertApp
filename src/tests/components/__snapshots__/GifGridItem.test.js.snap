// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en GifGridItem Should show the component correctly 1`] = `
<div
  className="card animate__animated animate__fadeIn"
>
   
  <img
    alt="Un título"
    src="https://localhost/img.jpg"
  />
  <p>
    Un título
  </p>
</div>
`;
